provider/spark.go

```diff
@@ -626,8 +630,172 @@ func (e EMRExecutor) InitializeExecutor(store FileStore) error {
+	sparkScriptPath := helpers.GetEnv("SPARK_SCRIPT_PATH", "/scripts/offline_store_spark_runner.py")
+	sparkScriptPathWithPrefix := store.PathWithPrefix(sparkScriptPath, false)
+
+	err := readAndUploadFile(sparkScriptPath, sparkScriptPathWithPrefix, store)
+	scriptExists, _ := store.Exists(sparkScriptPathWithPrefix)
+	if err != nil && !scriptExists {
+		return fmt.Errorf("could not upload spark script: Path: %s, Error: %v", sparkScriptPathWithPrefix, err)
+	}
+	return nil
+}
+
+func (s *SparkGenericExecutor) RunSparkJob(args *[]string, store FileStore) error {
+	bashCommand := "bash"
+	sparkArgs := *args
+	sparkArgsString := strings.Join(sparkArgs, " ")
+	bashCommandArgs := []string{"-c", fmt.Sprintf("pyenv global %s && pyenv exec %s", s.pythonVersion, sparkArgsString)}
 
+	s.logger.Infow("Executing the", bashCommand, "command")
+	cmd := exec.Command(bashCommand, bashCommandArgs...)
+	cmd.Env = append(os.Environ(), "FEATUREFORM_LOCAL_MODE=true")
+
+	err := cmd.Start()
+	if err != nil {
+		return fmt.Errorf("could not run spark job: %v", err)
+	}
+
+	err = cmd.Wait()
+	if err != nil {
+		return fmt.Errorf("spark job failed: %v", err)
+	}
+
+	return nil
+}
+
+func (s *SparkGenericExecutor) PythonFileURI(store FileStore) string {
+	return ""
+}
+
+func (s *SparkGenericExecutor) SparkSubmitArgs(destPath string, cleanQuery string, sourceList []string, jobType JobType, store FileStore) []string {
+	sparkScriptPath := helpers.GetEnv("SPARK_SCRIPT_PATH", "/scripts/offline_store_spark_runner.py")
+
+	argList := []string{
+		"spark-submit",
+		"--master",
+		s.master,
+		"--deploy-mode",
+		s.deployMode,
+		"--packages",
+		"\"org.apache.hadoop:hadoop-azure:3.2.0\"",
+		sparkScriptPath,
+		"sql",
+		"--output_uri",
+		fmt.Sprintf("\"%s\"", destPath),
+		"--sql_query",
+		fmt.Sprintf("\"%s\"", cleanQuery),
+		"--job_type",
+		fmt.Sprintf("\"%s\"", string(jobType)),
+	}
+
+	var remoteConnectionArgs []string
+	azureStore := store.AsAzureStore()
+	if azureStore != nil {
+		remoteConnectionArgs = []string{
+			"--spark_config",
+			fmt.Sprintf("\"%s\"", azureStore.configString()),
+		}
+	}
+	argList = append(argList, remoteConnectionArgs...)
+
+	argList = append(argList, "--source_list")
+	for _, source := range sourceList {
+		argList = append(argList, fmt.Sprintf("\"%s\"", source))
+	}
+	return argList
+}
+
+func (s *SparkGenericExecutor) GetDFArgs(outputURI string, code string, sources []string, store FileStore) ([]string, error) {
+	sparkScriptPath := helpers.GetEnv("SPARK_SCRIPT_PATH", "/scripts/offline_store_spark_runner.py")
+
+	argList := []string{
+		"spark-submit",
+		"--master",
+		s.master,
+		"--deploy-mode",
+		s.deployMode,
+		"--packages",
+		"\"org.apache.hadoop:hadoop-azure:3.2.0\"",
+		sparkScriptPath,
+		"df",
+		"--output_uri",
+		fmt.Sprintf("\"%s\"", outputURI),
+		"--code",
+		code,
+	}
+
+	var remoteConnectionArgs []string
+	azureStore := store.AsAzureStore()
+
+	if azureStore != nil {
```

Comment:
What is this doing?

Reply:
checks to see if azureStore in order to add credentials for it which will be used to pull the df pickle.

Reply:
Could we write the code to make this cleaner?
Maybe add an IsAzureStore method?

Reply:
Can we abstract this into its own interface of struct.

Example: have a new struct called
SparkArgs
- SubmitArgs
- DFArgs

Have two fns:
genericSparkArgs(...)
azureSparkArgs(...)

Then you can do a type switch on the store here.

You don't have to do this, but I personally think this would be much cleaner and extensible that what you did here.